c++/src/capnp/compat/websocket-rpc.c++

```diff
@@ -0,0 +1,107 @@
+#include <capnp/compat/websocket-rpc.h>
+#include <kj/io.h>
+#include <capnp/serialize.h>
+
+namespace capnp {
+
+WebSocketMessageStream::WebSocketMessageStream(kj::WebSocket& socket, size_t maxMessageReceiveSize)
+  : socket(socket),
+    maxMessageReceiveSize(maxMessageReceiveSize)
+  {};
+
+kj::Promise<kj::Maybe<MessageReaderAndFds>> WebSocketMessageStream::tryReadMessage(
+    kj::ArrayPtr<kj::AutoCloseFd> fdSpace,
+    ReaderOptions options, kj::ArrayPtr<word> scratchSpace) {
+  return socket.receive(maxMessageReceiveSize)
+    .then([options, scratchSpace](auto msg) -> kj::Promise<kj::Maybe<MessageReaderAndFds>> {
+      KJ_SWITCH_ONEOF(msg) {
+        KJ_CASE_ONEOF(closeMsg, kj::WebSocket::Close) {
+          return kj::Maybe<MessageReaderAndFds>();
+        }
+        KJ_CASE_ONEOF(str, kj::String) {
+          KJ_FAIL_REQUIRE(
+            "Unexpected websocket text message; expected only binary messages."
+          );
+          break;
+        }
+        KJ_CASE_ONEOF(bytes, kj::Array<byte>) {
+          kj::ArrayPtr<byte> ptr(bytes);
+          if(reinterpret_cast<uintptr_t>(&bytes[0]) % alignof(word) == 0
+              && bytes.size() % sizeof(word) == 0) {
+            return kj::Maybe(MessageReaderAndFds {
+              .reader = kj::heap<FlatArrayMessageReader>(
+                  kj::arrayPtr(
+                    reinterpret_cast<word *>(&bytes[0]),
+                    bytes.size() / sizeof(word)
+                  ),
+                  options
+                ).attach(kj::mv(bytes)),
+              .fds = nullptr
+            });
+          } else {
+            // The array is misaligned or not a whole number of words, so we
+            // need to copy it.
+            auto stream = kj::heap<kj::ArrayInputStream>(bytes);
+            auto reader = kj::heap<InputStreamMessageReader>(*stream, options, scratchSpace);
```

Comment:
I think it would be cheaper to make an aligned copy of the whole buffer rather than to use `InputStreamMessageReader` around `ArrayInputStream`. These classes will end up copying all the underlying bytes anyway and will do more allocations overall.